xray/sql_context.go

```diff
@@ -0,0 +1,752 @@
+		}
+	}
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	// execute query
+	var rows driver.Rows
+	if queryCtx, ok := stmt.(driver.StmtQueryContext); ok {
+		rows, err = queryCtx.QueryContext(ctx, []driver.NamedValue{})
+	} else {
+		select {
+		default:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		rows, err = stmt.Query([]driver.Value{})
+	}
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	// scan
+	if len(dest) != len(rows.Columns()) {
+		return fmt.Errorf("xray: expected %d destination arguments in Scan, not %d", len(rows.Columns()), len(dest))
+	}
+	cols := make([]driver.Value, len(rows.Columns()))
+	if err := rows.Next(cols); err != nil {
+		return err
+	}
+	for i, src := range cols {
+		d := dest[i]
+		switch s := src.(type) {
+		case string:
+			*d = s
+		case []byte:
+			*d = string(s)
+		case time.Time:
+			*d = s.Format(time.RFC3339Nano)
+		case int64:
+			*d = strconv.FormatInt(s, 10)
+		case float64:
+			*d = strconv.FormatFloat(s, 'g', -1, 64)
+		case bool:
+			*d = strconv.FormatBool(s)
+		default:
+			return fmt.Errorf("sql: Scan error on column index %d, name %q: type missmatch", i, rows.Columns()[i])
+		}
+	}
+
+	return nil
+}
+
+func (attr *dbAttribute) populate(ctx context.Context, query string) {
+	seg := GetSegment(ctx)
+
+	if seg == nil {
+		processNilSegment(ctx)
+		return
+	}
+
+	seg.Lock()
+	seg.Namespace = "remote"
+	seg.GetSQL().ConnectionString = attr.connectionString
+	seg.GetSQL().URL = attr.url
+	seg.GetSQL().DatabaseType = attr.databaseType
+	seg.GetSQL().DatabaseVersion = attr.databaseVersion
+	seg.GetSQL().DriverVersion = attr.driverVersion
+	seg.GetSQL().User = attr.user
+	seg.GetSQL().SanitizedQuery = query
+	seg.Unlock()
+}
+
+type driverTx struct {
+	driver.Tx
+}
+
+func (tx *driverTx) Commit() error {
+	return tx.Tx.Commit()
+}
+
+func (tx *driverTx) Rollback() error {
+	return tx.Tx.Rollback()
+}
+
+type driverStmt struct {
+	driver.Stmt
+	conn  *driverConn
+	attr  *dbAttribute
+	query string
+}
+
+func (stmt *driverStmt) Close() error {
+	return stmt.Stmt.Close()
+}
+
+func (stmt *driverStmt) NumInput() int {
+	return stmt.Stmt.NumInput()
+}
+
+func (stmt *driverStmt) Exec(args []driver.Value) (driver.Result, error) {
+	panic("not supported")
+}
+
+func (stmt *driverStmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	var result driver.Result
+	var err error
+	if execerContext, ok := stmt.Stmt.(driver.StmtExecContext); ok {
+		err = Capture(ctx, stmt.attr.dbname, func(ctx context.Context) error {
+			stmt.populate(ctx)
+			var err error
+			result, err = execerContext.ExecContext(ctx, args)
+			return err
+		})
+	} else {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+		dargs, err0 := namedValuesToValues(args)
+		if err0 != nil {
+			return nil, err0
+		}
+		err = Capture(ctx, stmt.attr.dbname, func(ctx context.Context) error {
+			stmt.populate(ctx)
+			var err error
+			result, err = stmt.Stmt.Exec(dargs)
```

Comment:
can we disable deprecated warning reported by golangci-lint ?

Reply:
ignore warnings and merge this pull request please.
[reviewdog](https://github.com/reviewdog/reviewdog) will not report them next time.

Reply:
Hi @shogo82148,

I am going to merge this PR today. Just wanted to provide final suggestions. In one of the subsegment Sanitized Query field looks like this: 

Sanitized query | SELECT bar FROM foo WHERE bar = ? -- skipped

Do we want to add more information so that customer can understand why we get subsegment of skipped query and the same query gets traced successfully in next subsegment  ?

Reply:
I have just committed https://github.com/aws/aws-xray-sdk-go/pull/169/commits/328f078137c50bb6735b34b16ed7666914af8873
how about it?

it comes from the message of `driver.ErrSkip`

```go
var ErrSkip = errors.New("driver: skip fast-path; continue as if unimplemented")
```

Reply:
Sorry for last min suggestions. I was wondering what is the point of tracing this error skip message as subsegments ? I think tracing this message will result in 2 subsegments of the same query which is redundant. Moreover, this does not add any instrumentation value as this msg just informs sql package should continue as if the optional interface is not implemented. What do you think ?

Reply:
Yes, you are right. This subsegment may be redundant.
But there is no interface to remove existing subsegments, and we can't know whether we should add a new subsegment before calling `Exec`.

And the users can use this message as optimize hint.
e.g. the [interpolateParams](https://github.com/go-sql-driver/mysql#interpolateparams) option of go-sql-driver/mysql enables this fast-path. The users can know whether this option really works well via the trace.

Reply:
Okay Makes sense! Also, there was a customer ask to keep sql.go 
(previous implementation) and add a note that older (sql.go) implementation will be deprecated when we do Go-SDK GA. This will give more time to customers who use the existing implementation. Would you mind adding deleted file so when I merge this PR it will have the previous implementation same as the base branch.

Reply:
Okay. that's fair!